api/server/response_options.proto

```diff
@@ -5,6 +5,15 @@ package nighthawk.server;
 import "google/protobuf/wrappers.proto";
 import "validate/validate.proto";
 import "envoy/api/v2/core/base.proto";
+import "google/protobuf/duration.proto";
+
+message GaugeBasedDelay {
+  // Name of the gauge providing the target value.
+  string gauge_name = 1;
+  int64 gauge_target_value = 2;
+  // The delay is computed as (guauge_target_value - gauge_value) * delta_delay;
```

Comment:
Considering a use case where we want to inject latency proportional to the in-flight request count, how would we implement that using this config?

It looks like this models a gauge that starts as a positive value and decreases toward zero, but a gauge like downstream_rq_active does the opposite. If negative values for delta_delay or gauge_target_value are supported, could you state that explicitly?

Also, I think we need the first request to incur some latency, otherwise it will take pretty severe load before the in-flight count grows at all. This tells me that the model needs to allow an offset value - i.e. a baseline latency when the gauge is at the target.

Reply:
To put it another way: right now I think you're proposing an equation like y = max(0, m * (x_0 - x)) with unspecified constraints on the RHS variables. I think it would be easier to reason about y = max(0, m * x + b) where m and b can be any real values.

Reply:
@jc1e100 Currently the delay linearly shrinks to zero as the gauge approaches the target.
This is expressed as follows in the code:

```c++
    // Substracts one from the gauge value to avoid inclusion of the the current request
    delay = (gauge_target_value - (gauge_value - 1)) * delay_delta;
    if (delay <= 0) {
      delay = absl::nullopt;
    }
```

What went wrong here is that I misunderstood, and thought we had been aiming to get a steady number of in-flight downstream requests.

I'll make changes based on your input as well as Josh his earlier comment to simplify this.

Reply:
The delay now looks like below (any negative computed delays are ignored / not propagated):

```
    auto& concurrency = base_config_.concurrency_based_delay();
    const uint64_t current_value = config_->approximateInstances();
    delay = Envoy::Protobuf::util::TimeUtil::DurationToMilliseconds(
        concurrency.minimal_delay() + (current_value * concurrency.concurrency_delay_factor()));
```